### PR TITLE
Test: Added init-method property test

### LINKED ISSUE #1471: init
The generated canister will look something like this:

```ts
// Create variables for each init param 
let initParam0;
let initParam1;

export default Canister {
    init: ([bool, text], (param0) => {
        // check that they are the right order/length
        // store the params in the global variables 
        initParam0 = param0;
        initParam1 = param1;
    }),
    // create a method to retrieve the stored values
    getInitValues: query([], Tuple(bool,text), () => {
        return [initParam0, initParam1]
    })
}
```

The Tests should also run with 0-20 query and update methods.

You'll also need to figure out how to pass params to the deploy command.

Subtasks
- [x] Support an `init` method with 0 parameters
- [x] Support multiple params passed across the CLI to dfx deploy
  - [x] Update CandidValueAndMeta, combining agentArgumentValue, agentResponseValue into a new values object
  - [x] Add a CandidTypeObject to the new values object in CandidValueAndMeta
  - [x] Collect the params and call valueToString and pass them to dfx deploy
  - [x] Figure out what's wrong with float32s and get seeds `-2020534984` and `-210693065` to pass.

### PR DESCRIPTION
**Pull request by dansteren**
Closes #1471

### CODE REVIEW

File: property_tests/arbitraries/canister_methods/init_method_arb.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        >;
+        callbackLocation?: CallbackLocation;
+    }
+) {
+    return fc
+        .tuple(
+            UniqueIdentifierArb('canisterMethod'),
+            paramTypeArrayArb,
+            VoidArb(),
+            fc.constantFrom(
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply property_tests/arbitraries/canister_methods/init_method_arb.ts:54]: Pass in CallbackLocation as a type argument

File: property_tests/arbitraries/canister_methods/init_method_arb.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            // that for now.
+        )
+        .map(
+            ([
+                functionName,
+                paramTypes,
+                returnType,
+                defaultCallbackLocation,
+                callbackName
+            ]) => {
```
Comment (lastmjs) [inline property_tests/arbitraries/canister_methods/init_method_arb.ts:70]: Resolve this TODO

File: property_tests/arbitraries/candid/simple_type_arbs/definition_arb.ts [hunk 3]
```diff
@@ -22,6 +24,7 @@ export function SimpleCandidDefinitionArb(
                     candidTypeAnnotation,
                     candidTypeObject,
                     candidType,
+                    azleCandidTypeObject,
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/simple_type_arbs/definition_arb.ts:27]: Rename `azleCandidTypeObject` to `runtimeCandidTypeObject`, rename this everywhere appropriate

File: property_tests/arbitraries/candid/simple_type_arbs/candid_type_to_azle_candid_type.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    nat64,
+    Null,
+    text,
+    Void,
+    Principal,
+    CandidType
+} from '../../../../src/lib';
+import { SimpleCandidType } from '../candid_type';
+
+export function candidTypeToAzleCandidType(
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/simple_type_arbs/candid_type_to_azle_candid_type.ts:24]: Rename to `candidTypeToRuntimeCandidTypeObject`

File: property_tests/arbitraries/candid/reference/service_arb/definition_arb.ts [hunk 5]
```diff
@@ -148,6 +110,19 @@ function generateCandidTypeObject(
     return `Canister({${methods}})`;
 }
 
+function generateAzleCandidTypeObject(
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/reference/service_arb/definition_arb.ts:113]: Rename to `generateRuntimeCandidTypeObject`

File: property_tests/arbitraries/candid/reference/service_arb/service_method_arb.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            };
+        });
+}
+
+function generateAzleCandidTypeObject(
+    mode: Mode,
+    params: CandidDefinition[],
+    returnType: CandidDefinition
+) {
+    const modeFunction = mode === 'query' ? query : update;
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/reference/service_arb/service_method_arb.ts:78]: Rename

File: property_tests/arbitraries/candid/constructed/record_arb/definition_arb.ts [hunk 7]
```diff
@@ -97,6 +101,17 @@ function generateCandidTypeObject(
         .join(',')}})`;
 }
 
+function generateAzleCandidTypeObject(fields: Field[]): CandidType {
+    const azleRecordConstructorObj = fields.reduce<{
+        [key: string]: CandidType;
+    }>((acc, [fieldName, fieldDefinition]) => {
+        acc[fieldName] = fieldDefinition.candidMeta.azleCandidTypeObject;
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/constructed/record_arb/definition_arb.ts:108]: No mutations

File: property_tests/arbitraries/candid/constructed/record_arb/definition_arb.ts [hunk 8]
```diff
@@ -97,6 +101,17 @@ function generateCandidTypeObject(
         .join(',')}})`;
 }
 
+function generateAzleCandidTypeObject(fields: Field[]): CandidType {
+    const azleRecordConstructorObj = fields.reduce<{
+        [key: string]: CandidType;
+    }>((acc, [fieldName, fieldDefinition]) => {
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/constructed/record_arb/definition_arb.ts:107]: Remove type argument, statically type params and return type if necessary

File: property_tests/arbitraries/candid/constructed/variant_arb/definition_arbs.ts [hunk 9]
```diff
@@ -120,3 +124,14 @@ function generateCandidTypeObject(
         )
         .join(',')}})`;
 }
+
+function generateAzleCandidTypeObject(fields: Field[]): CandidType {
+    const azleVariantConstructorObj = fields.reduce<{
+        [key: string]: CandidType;
+    }>((acc, [fieldName, fieldDefinition]) => {
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/constructed/variant_arb/definition_arbs.ts:131]: Static type the reduce method itself, get rid of the type argument, no mutations

File: property_tests/arbitraries/candid/constructed/variant_arb/values_arb.ts [hunk 10]
```diff
@@ -26,10 +26,10 @@ export function VariantValuesArb(
         return [name, values];
     });
 
-    return fieldValue.map((fieldValues) => {
-        const valueLiteral = generateValueLiteral(fieldValues);
-        const agentArgumentValue = generateValue(fieldValues);
-        const agentResponseValue = generateValue(fieldValues, true);
+    return fieldValue.map((field) => {
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/constructed/variant_arb/values_arb.ts:29]: Rename `fieldValue` to `fieldValues`, rename `field` to `fieldValue`.

File: property_tests/arbitraries/candid/candid_value_and_meta_arb.ts [hunk 11]
```diff
@@ -71,3 +75,31 @@ export function CandidValueAndMetaArb(): fc.Arbitrary<
 }
 
 // TODO: This needs to support service.
+
+/**
+ * Azle currently (v0.18.6) cannot accept funcs in init methods.
+ * See https://github.com/demergent-labs/azle/issues/1474
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/candid_value_and_meta_arb.ts:81]: Resolve this issue

File: property_tests/arbitraries/candid/candid_value_and_meta_arb_generator.ts [hunk 12]
```diff
@@ -34,8 +35,11 @@ export function CandidValueAndMetaArbGenerator<
                     imports,
                     valueLiteral
                 },
-                agentArgumentValue,
-                agentResponseValue
+                value: {
+                    agentArgumentValue,
+                    agentResponseValue,
+                    candidTypeObject: azleCandidTypeObject
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/candid_value_and_meta_arb_generator.ts:41]: Deal with the naming here, seems like both should be named `runtimeCandidTypeObject`.

File: property_tests/arbitraries/candid/candid_definition_arb/types.ts [hunk 13]
```diff
@@ -59,16 +61,11 @@ export type ServiceCandidDefinition = {
...
-    name: string;
-    imports: Set<string>;
-    variableAliasDeclarations: string[];
-    src: string;
-};
 
 type CandidMeta = {
     candidTypeAnnotation: string; // Either a type reference or type literal
     candidTypeObject: string;
+    azleCandidTypeObject: AzleCandidType;
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/candid_definition_arb/types.ts:68]: Rename azleCandidTypeObject, AzleCandidType type seems like it has a weird name or type...

File: property_tests/tests/init_method/package.json [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{
+    "scripts": {
+        "test": "ts-node --transpile-only --ignore=false test/test.ts"
+    },
+    "dependencies": {
+        "azle": "^0.18.5"
```
Comment (lastmjs) [inline property_tests/tests/init_method/package.json:6]: Add to the deployment script to change these versions

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: src/lib/candid/did_file/visitor/visit/service.ts [hunk 15]
```diff
@@ -125,7 +125,7 @@ function createCanisterParamsString(
     }
     const parts = initMethodCandidString[0].split('->');
     if (parts.length >= 2) {
-        return parts[0].trim();
+        return parts.slice(0, -1).join('->').trim();
```
Comment (lastmjs) [inline src/lib/candid/did_file/visitor/visit/service.ts:128]: I just have a question about this

File: property_tests/arbitraries/candid/candid_value_and_meta_arb_generator.ts [hunk 16]
```diff
@@ -10,17 +10,21 @@ export function CandidValueAndMetaArbGenerator<
     V extends CandidValues<T>
 >(
     DefinitionArb: fc.Arbitrary<D>,
-    ValueArb: (arb: D) => fc.Arbitrary<V>
+    ValueArb: (arb: D, constraints?: any) => fc.Arbitrary<V>,
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/candid_value_and_meta_arb_generator.ts:13]: I have a question about this

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: property_tests/tests/init_method/package.json (hunk 14 above)
Comment (bdemann) [reply property_tests/tests/init_method/package.json:6]: it is now

**Review by lastmjs — APPROVED** [review]

**Review by lastmjs — APPROVED** [review]

**Pull request merged**